File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/python/h4ckxel.py
"""
Operadores 
"""

# Aritmeticos 
print(f"Suma: 10 + 8 = {10 + 8}")
print(f"Resta: 10 - 8 = {10 - 8}")
print(f"Multiplicacion: 10 * 8 = {10 * 8}")
print(f"Division: 10 / 8 = {10 / 8}")
print(f"Modulo: 10 % 8 = {10 % 8}")
print(f"Exponente: 10 ** 0 = {10 ** 8}")
print(f"División entera: 10 // 8 = {10 // 8}")

# Comparacion 
print(f"igualdad: 10 == 3 es {10 == 3}")
print(f"Desigualdad: 10 != 3 es {10 != 3}")
print(f"Mayor que: 10 > 3 es {10 > 3}")
print(f"Menor que: 10 < 3 es {10 < 3}")
print(f"Mayor o igual que: 10 >= 10 es {10 >= 10}")
print(f"Menor o igual que: 10 <= 3 es {10 <= 3}")

# Lógicos
print(f"AND &&: 10 + 3 == 13 and 5 - 1 == 4 es {10 + 3 == 13 and 5 -1 == 4}")
print(f"OR ||: 10 + 3 == 13 or 5 - 1 == 4 es {10 + 3 == 14 or 5 - 1 == 4}")
print(f"NOT !: not 10 + 3 == 14 es {not 10 + 3 == 14}")

# Asignacón 
my_number = 11  
print(my_number)
my_number += 1  
print(my_number)
my_number -= 1  
print(my_number)
my_number *= 2  
print(my_number)
my_number /= 2  
print(my_number)
my_number %= 2  
print(my_number)
my_number **= 1  
print(my_number)
my_number //= 1  
print(my_number)

# Identida 
my_new_number = my_number
print(f"my_number is my_new_number es {my_number is my_new_number}")
print(f"my_number is not my_new_number es {my_number is not my_new_number}")

# Pertenencia 
print(f"'y' in 'Python' = {'y' in 'Python'}")
print(f"'b' not in 'Python' = {'b' not in 'Python'}")

# Bit 
a = 10  # 1010
b = 3  # 0011
print(f"AND: 10 & 3 = {10 & 3}")  # 0010
print(f"OR: 10 | 3 = {10 | 3}")  # 1011
print(f"XOR: 10 ^ 3 = {10 ^ 3}")  # 1001
print(f"NOT: ~10 = {~10}")
print(f"Desplazamiento a la derecha: 10 >> 2 = {10 >> 2}")  # 0010
print(f"Desplazamiento a la izquierda: 10 << 2 = {10 << 2}")  # 101000

"""
Estructuras de control
"""
# Condicionales

my_string = "Python"

if my_string == "Rust":
    print("my_string es 'Rust'")
elif my_string == "Python":
    print("my_string es 'Python'")
else:
    print("my_string no es 'Rust' ni 'Python'")

# Iterativas

for i in range(11):
    print(i)

i = 0

while i <= 10:
    print(i)
    i += 1

# Manejo de excepciones
try:
    print(10 / 0)
except:
    print("Se ha producido un error")
finally:
    print("Ha finalizado el manejo de excepciones")

"""
Extra
"""

for number in range(10, 56):
    if number % 2 == 0 and number != 16 and number % 3 != 0:
        print(number)

File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/javascript/h4ckxel.js

// Operadores Aritméticos

let a = 10;
let b = 5;

console.log("Suma:", a + b);         // 15
console.log("Resta:", a - b);        // 5
console.log("Multiplicación:", a * b); // 50
console.log("División:", a / b);     // 2
console.log("Módulo:", a % b);       // 0
console.log("Exponenciación:", a ** b); // 100000







// Operadores Lógicos

let x = true;
let y = false;

console.log("AND lógico:", x && y); // false
console.log("OR lógico:", x || y);  // true
console.log("NOT lógico:", !x);     // false






// Operadores de Comparación
let p = 10;
let q = 20;

console.log("Igual a:", p == q);        // false
console.log("No igual a:", p != q);     // true
console.log("Mayor que:", p > q);       // false
console.log("Menor que:", p < q);       // true
console.log("Mayor o igual que:", p >= q); // false
console.log("Menor o igual que:", p <= q); // true







// Operadores de Asignación
let n = 5;

n += 3; // n = n + 3
console.log("Asignación de suma:", n); // 8

n -= 2; // n = n - 2
console.log("Asignación de resta:", n); // 6

n *= 4; // n = n * 4
console.log("Asignación de multiplicación:", n); // 24

n /= 3; // n = n / 3
console.log("Asignación de división:", n); // 8

n %= 2; // n = n % 2
console.log("Asignación de módulo:", n); // 0








// Operadores de Identidad

let r = 5;
let s = "5";

console.log("Identidad (igual):", r === s); // false
console.log("Identidad (no igual):", r !== s); // true









// Operadores de Pertenencia

let arr = [1, 2, 3, 4, 5];

console.log("Pertenencia (en array):", arr.includes(3)); // true
console.log("Pertenencia (no en array):", arr.includes(6)); // false







// Operadores de Bits

let m = 5;  // 0101 en binario
let ñ = 3;  // 0011 en binario

console.log("AND bit a bit:", m & ñ); // 1 (0001 en binario)
console.log("OR bit a bit:", m | ñ);  // 7 (0111 en binario)
console.log("XOR bit a bit:", m ^ ñ); // 6 (0110 en binario)
console.log("Complemento bit a bit:", ~m); // -6 (invertido en binario)
console.log("Desplazamiento a la izquierda:", m << 1); // 10 (1010 en binario)
console.log("Desplazamiento a la derecha:", m >> 1); // 2 (0010 en binario)
















// Condicionales

let age = 20;

if (age < 18) {
    console.log("Eres menor de edad.");
} else if (age >= 18 && age < 65) {
    console.log("Eres adulto.");
} else {
    console.log("Eres mayor de edad.");
}





// Iterativas


// Bucle for
for (let i = 0; i < 5; i++) {
    console.log("Bucle for, iteración:", i);
}

// Bucle while
let j = 0;
while (j < 5) {
    console.log("Bucle while, iteración:", j);
    j++;
}

// Bucle do-while
let k = 0;
do {
    console.log("Bucle do-while, iteración:", k);
    k++;
} while (k < 5);





// Excepciones

try {
    let result = 10 / 0;
    if (!isFinite(result)) {
        throw new Error("División por cero");
    }
    console.log("Resultado:", result);
} catch (error) {
    console.log("Error capturado:", error.message);
} finally {
    console.log("Bloque finally ejecutado.");
}









// * DIFICULTAD EXTRA (opcional):
//   * Crea un programa que imprima por consola todos los números comprendidos
//   * entre 10 y 55 (incluidos), pares, y que no son ni el 16 ni múltiplos de 3.
//   *
//   * Seguro que al revisar detenidamente las posibilidades has descubierto algo nuevo.

    
function imprimir_numeros(){
    for ( let x = 10; x < 56; x++){
        if (x >= 10 && x <=55 && x % 2 === 0 && x != 16 && x % 3 != 0) {
            console.log(x)
        }
    }
    
}
imprimir_numeros()

